Method as function

The following program `sample.go`:

```go
package main

import (
	"fmt"
)

type Cheese struct {
	property string
}

func (t *Cheese) Hello(param string) {
	fmt.Printf("%+v %+v", t, param)
}

func main() {
	(*Cheese).Hello(&Cheese{property: "value"}, "param")
}

// Output:
// &{property:value} param
```

Expected result:
```console
$ go run ./sample.go
&{property:value} param
```

Got:
```console
$ dyngo ./sample.go
_test/receiver.go:16:2: undefined type
```

--- a/_test/method16.go
+++ b/_test/method16.go
@@ -1,0 +1,20 @@
+package main
+
+import (
+	"fmt"
+)
+
+type Cheese struct {
+	property string
+}
+
+func (t *Cheese) Hello(param string) {
+	fmt.Printf("%+v %+v", t, param)
+}
+
+func main() {
+	(*Cheese).Hello(&Cheese{property: "value"}, "param")
+}
+
+// Output:
+// &{Xproperty:value} param

--- a/interp/cfg.go
+++ b/interp/cfg.go
@@ -1004,11 +1004,21 @@ func (interp *Interpreter) Cfg(root *Node) ([]*Node, error) {
 					err = n.cfgError("undefined selector: %s", n.child[1].ident)
 				}
 			} else if m, lind := n.typ.lookupMethod(n.child[1].ident); m != nil {
-				// Handle method
-				n.gen = getMethod
-				n.val = m
-				n.typ = m.typ
-				n.recv = &Receiver{node: n.child[0], index: lind}
+				if n.child[0].isType(scope) {
+					// Handle method as a function with receiver in 1st argument
+					n.val = m
+					n.findex = -1
+					n.gen = nop
+					n.typ = &Type{}
+					*n.typ = *m.typ
+					n.typ.arg = append([]*Type{n.child[0].typ}, m.typ.arg...)
+				} else {
+					// Handle method with receiver
+					n.gen = getMethod
+					n.val = m
+					n.typ = m.typ
+					n.recv = &Receiver{node: n.child[0], index: lind}
+				}
 			} else if ti := n.typ.lookupField(n.child[1].ident); len(ti) > 0 {
 				// Handle struct field
 				n.val = ti
@@ -1053,6 +1063,10 @@ func (interp *Interpreter) Cfg(root *Node) ([]*Node, error) {
 			case n.anc.kind == Field:
 				// pointer type expression in a field expression (arg or struct field)
 				n.gen = nop
+			case n.anc.kind == ParenExpr && n.child[0].sym != nil && n.child[0].sym.kind == Typ:
+				// pointer type expression
+				n.gen = nop
+				n.typ = &Type{cat: PtrT, val: n.child[0].sym.typ}
 			default:
 				// dereference expression
 				wireChild(n)
@@ -1244,6 +1258,10 @@ func (n *Node) isType(scope *Scope) bool {
 	switch n.kind {
 	case ArrayType, ChanType, FuncType, MapType, StructType, Rtype:
 		return true
+	case ParenExpr, StarExpr:
+		if len(n.child) == 1 {
+			return n.child[0].isType(scope)
+		}
 	case SelectorExpr:
 		pkg, name := n.child[0].ident, n.child[1].ident
 		if sym, _, ok := scope.lookup(pkg); ok {

--- a/interp/interp_consistent_test.go
+++ b/interp/interp_consistent_test.go
@@ -33,6 +33,7 @@ func TestInterpConsistencyBuild(t *testing.T) {
 			file.Name() == "io0.go" || // use random number
 			file.Name() == "op1.go" || // expect error
 			file.Name() == "bltn0.go" || // expect error
+			file.Name() == "method16.go" || // private struct field
 			file.Name() == "switch8.go" || // expect error
 			file.Name() == "switch9.go" || // expect error
 			file.Name() == "switch13.go" || // expect error

--- a/interp/interp_test.go
+++ b/interp/interp_test.go
@@ -3756,6 +3756,37 @@ func main() {
 	// 25
 }
 
+func Example_method16() {
+	src := `
+package main
+
+import (
+	"fmt"
+)
+
+type Cheese struct {
+	property string
+}
+
+func (t *Cheese) Hello(param string) {
+	fmt.Printf("%+v %+v", t, param)
+}
+
+func main() {
+	(*Cheese).Hello(&Cheese{property: "value"}, "param")
+}
+`
+	i := interp.New(interp.Opt{Entry: "main"})
+	i.Use(stdlib.Value)
+	_, err := i.Eval(src)
+	if err != nil {
+		panic(err)
+	}
+
+	// Output:
+	// &{Xproperty:value} param
+}
+
 func Example_method2() {
 	src := `
 package main

